feat: add public key signature verification support

ahmad-ibra opened the pull request:
Adds support for public key signature verification of artifacts signed with `cosign`

TylerGillson commented on api/v1alpha1/ocivalidator_types.go:
```diff
@@ -70,6 +76,16 @@ type Auth struct {
+	// +kubebuilder:default:=cosign
+	Provider string `json:"provider" yaml:"provider"`
+
+	// SecretName is the name of the Kubernetes Secret that contains trusted public keys used to sign artifacts in the OciRegistryRule
```
It would be useful to mention here that the secret referred to by SecretName must exist in the same namespace as the OCIValidator.

TylerGillson commented on api/v1alpha1/ocivalidator_types.go:
```diff
@@ -70,6 +76,16 @@ type Auth struct {
 	SecretName string `json:"secretName" yaml:"secretName"`
```
It would be useful to mention here that the auth secret referred to by SecretName must exist in the same namespace as the OCIValidator.

TylerGillson commented on internal/controller/ocivalidator_controller.go:
```diff
@@ -105,33 +107,67 @@ func (r *OciValidatorReconciler) secretKeyAuth(req ctrl.Request, rule v1alpha1.O
+	}
+
+	for k, data := range pubKeysSecret.Data {
+		// search fro public keys in the secret
```
nit: fro -> for

TylerGillson commented on internal/validators/oci_validator.go:
```diff
@@ -59,10 +65,15 @@ func (s *OciRuleService) ReconcileOciRegistryRule(rule v1alpha1.OciRegistryRule,
 	errs := make([]error, 0)
 	details := make([]string, 0)
+
+	if pubKeys != nil && len(pubKeys) == 0 {
```
This nil check is redundant. We should be seeing a lint error here for a tautological condition. Maybe we need to enable some more golangci-lint linters.

ahmad-ibra replied:
imo the nil check isnt redundant. for nil or empty slices `len(pubKeys)` will return 0, but an initialized and empty slice would not be nil.

example of this can be seen here:
https://go.dev/play/p/6c5H86p6131

what im really looking for is the case where a secret was provided and so we'd expect to see some secrets, but there were no public keys found within it.

TylerGillson replied:
you can use `len(pubKeys) > 0`

TylerGillson replied:
or, rather, `len(pubKeys) == 0` in this specific case. your example highlights that it'll work, since it outputs:
```
s1 is empty
s2 is empty
```

TylerGillson replied:
the length is zero in both bases

TylerGillson commented on internal/validators/oci_validator.go:
```diff
@@ -299,6 +316,34 @@ func setupTransportOpts(opts []remote.Option, caCert string) ([]remote.Option, e
+
+		verified, err := verifier.Verify(ctxTimeout, ref)
+		if err != nil {
+			continue
```
I'd recommend logging the error for debugging

ahmad-ibra replied:
Added details and errs in this case.

TylerGillson commented on internal/validators/oci_validator.go:
```diff
@@ -138,13 +149,19 @@ func validateReference(ref name.Reference, fullLayerValidation bool, opts []remo
 
 // validateRepos validates repos within a registry. This function is to be used when no particular artifact in a registry is provided
-func validateRepos(ctx context.Context, host string, opts []remote.Option, vr *types.ValidationResult) ([]string, error) {
+func validateRepos(ctx context.Context, host string, opts []remote.Option, pubKeys [][]byte, vr *types.ValidationResult) ([]string, error) {
```
You're not using the `ctx` param - you're creating a new one.

TylerGillson commented on internal/validators/oci_validator.go:
```diff
@@ -138,13 +149,19 @@ func validateReference(ref name.Reference, fullLayerValidation bool, opts []remo
 
-	// TODO: add signature verification here
+	if pubKeys != nil {
+		ctx := context.Background()
```
nit: consider passing the context from `ReconcileOciRegistryRule` into `validateReference` and forward it along here.

ahmad-ibra replied:
fixed up ctx usages all around since there were a few other issues i noticed

ahmad-ibra commented:
Addressed all the comments and audited the errors/details being added to the validation results. found a lot of cases where we werent too consistent with what was being added and made sure to address that.

Also found another area where i could bump our code coverage up pretty easily and made sure to batch that in with these changes

TylerGillson reviewed (approved):
lgtm
